src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c

```diff
@@ -153,42 +207,35 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
         return -1;
     }
 
-    size_t executable_path_len = strlen(bundle_path) + strlen(executable) + 2; // +2 for '/' and '\0'
-    char* executable_path = (char*)malloc(executable_path_len);
-    size_t res = sprintf (executable_path, "%s/%s", bundle_path, executable);
-    if (res != executable_path_len - 1)
-    {
-        LOG_ERROR("Failed to resolve full path for: %s", executable);
-        return -1;
-    }
-    executable_path[res] = '\0';
-
-    unsigned int coreclr_domainId = 0;
-    void *coreclr_handle = NULL;
-
     LOG_INFO ("Calling coreclr_initialize");
     int rv = coreclr_initialize (
-		executable_path,
+		g_executable_path,
 		executable,
 		3,
 		appctx_keys,
 		appctx_values,
-		&coreclr_handle,
-		&coreclr_domainId
+		&g_coreclr_handle,
+		&g_coreclr_domainId
 		);
     LOG_INFO ("coreclr_initialize returned %d", rv);
-
-    LOG_INFO ("Calling coreclr_execute_assembly");
-    coreclr_execute_assembly (coreclr_handle, coreclr_domainId, managed_argc, managed_argv, executable_path, &rv);
-
-    LOG_INFO ("Exit code: %d.", rv);
     return rv;
 }
 
+void
+Java_net_dot_MonoRunner_setEnv (JNIEnv* env, jobject thiz, jstring j_key, jstring j_value)
+{
+    LOG_INFO ("Java_net_dot_MonoRunner_setEnv:");
```

Comment:
I would assert that coreclr hasn't been initialized yet. In fact, it might make sense to return a non-zero value if someone calls this and coreclr has been initialized. The `setenv` is a tricky API and rather untrustworthy from a threading perspective.

Reply:
Why is it important that the runtime hasn't been initialized yet for setting env variables?

Reply:
My assumption here is that this is primarily for setting environment variables for use during coreclr initialization, that could be wrong. The issue with that is there are some environment variables that aren't obvious when they are read/consumed and it can cause confusion if all `DOTNET_` related variables are set a priori.

On a general note for `setenv`, it isn't thread safe. In coreclr we have a big lock around it for that very reason. However, we can't control other callers of that API and one can get into weird situation when reading the environment block and other threads are setting it. This could have been addressed by the Android system, but as far as I'm aware it is still a common issue with the API not being thread safe.